### ISSUE TITLE
Configuration option for excluding email addresses in exports

### ISSUE THREAD

**Issue by pameyer** [IQSS/dataverse#3443]
Email addresses, including datataset 'contact addresses', from users shouldn't be directly exposed (aka - open to external harvesting).

Thanks to @pdurbin for finding the first instance of this:

###
 Also, I just confirmed that "datasetContact" email addresses are available like so:

curl https://dataverse.harvard.edu/api/datasets/:persistentId?persistentId=doi:10.7910/DVN/26935 | jq '.data.latestVersion.metadataBlocks.citation.fields[] | select(.typeName=="datasetContact").value[0].datasetContactEmail.value'

We need to check with Pete to see if this constitutes leaking or not.

###

**Comment by djbrooke** [comment]
This is also exposed in DDI and JSON exports:

- https://dataverse.harvard.edu/api/datasets/export?exporter=dataverse_json&persistentId=doi%3A10.7910/DVN/26935
- https://dataverse.harvard.edu/api/datasets/export?exporter=ddi&persistentId=doi%3A10.7910/DVN/26935

**Comment by djbrooke** [comment]
It's recommended that an avenue for reaching the dataset contact is provided. Through the UI, this is provided by the "email dataset contact" functionality. In the metadata exports, this is provided by including the contact's email address.

Some installations may want to not expose email addresses in easily harvestable formats. 

At the installation level, provide a way for administrators to not make this information available.

**Comment by pdurbin** [comment]
I just made pull request #3791 and I'm putting this issue in Code Review at https://waffle.io/IQSS/dataverse

**Comment by pdurbin** [comment]
Changes to consider:

- Use FieldType = email for excluding rather than string comparison on the name.
- Change `walk(DatasetField fld, List<DatasetField> exclude)` to `walk(DatasetField fld, List<DatasetFieldType> exclude)`
- Instead of putting the exclude on the printer object... operate on FieldType instead. @scolapasta can clarify. Basically, can we prevent future exporters from introducing an regression on this issue, from exposing email addresses. 

To do:

- See how many FieldType = email instances there are. @scolapasta did a query and found only one.

**Comment by pdurbin** [comment]
@pameyer @kcondon I just found third API endpoint where the email address is exposed:

curl 'https://dataverse.harvard.edu/api/datasets/:persistentId/versions/:latest-published/metadata/citation?persistentId=doi:10.7910/DVN/26935' | jq '.data.fields[] | select(.typeName=="datasetContact").value[0].datasetContactEmail.value'

This is in addition to the two endpoints @djbrooke mentioned at https://github.com/IQSS/dataverse/issues/3443#issuecomment-282113235

**Comment by pdurbin** [comment]
> Change walk(DatasetField fld, List<DatasetField> exclude) to walk(DatasetField fld, List<DatasetFieldType> exclude)

Once I was in the code I realized that creating a list of fields to exclude doesn't make sense now that we're matching on `FieldType.EMAIL`. I was able to delete a lot of code by removing this exclude concept.

Since we are targeting an entire field type I could change `:ExcludeDatasetContactEmailFromExport` to `:ExcludeEmailFromExport`. I'm also open to other changes. I'm moving this to Code Review at https://waffle.io/IQSS/dataverse

The pull request is #3791.

**Comment by pdurbin** [comment]
I just opened a related issue that @landreev @sekmiller and I discussed this morning: Export/Harvesting: Leaving "Name" under "Contact" blank while creating a dataset results in no contact email address in the DDI export #3802 

I'm considering this out of scope for the present issue but I'd be happy to role in a fix for this if we so desire.

**Comment by pdurbin** [comment]
> I could change :ExcludeDatasetContactEmailFromExport to :ExcludeEmailFromExport

@scolapasta and I both like this change so I made it in dbde6df. Ready for code review.

**Comment by pdurbin** [comment]
Reviewed with @scolapasta and @sekmiller . To do:

- get tests working in DDIExporterTest

Also, think about this potential change:

```
diff --git a/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java b/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java
index ad65b08..e300f13 100644
--- a/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java
+++ b/src/main/java/edu/harvard/iq/dataverse/export/ExportService.java
@@ -33,7 +33,7 @@ public class ExportService {
     
     private static ExportService service;
     private ServiceLoader<Exporter> loader;
-    static SettingsServiceBean settingsService;
+    private SettingsServiceBean settingsService;
 
     private ExportService() {
         loader = ServiceLoader.load(Exporter.class);        
@@ -48,12 +48,13 @@ public class ExportService {
     }
 
     public static synchronized ExportService getInstance(SettingsServiceBean settingsService) {
-        ExportService.settingsService = settingsService;
+
         if (service == null) {
             service = new ExportService();
         } else{
             service.loader.reload();
         }
+        service.settingsService = settingsService;
         return service;
     }
```

**Comment by pdurbin** [comment]
To be clear, the issue is fixed from my perspective (and I've written tests to prove it 😄 ) but the todo list item about is desired from the perspective of not taking on technical debt. @sekmiller is going to see the null pointer exceptions when he changes this boolean to true: https://github.com/IQSS/dataverse/blob/dbde6dfe8cb67a468062d5965d78a05ade272952/src/test/java/edu/harvard/iq/dataverse/export/DDIExporterTest.java#L121

I put myself back on this issue to assist and talk about making that test real so we can delete some code elsewhere and still have code coverage of the DdiExportUtil class. It's complicated but I think @sekmiller @scolapasta and I are all on the same page.

**Comment by pdurbin** [comment]
> get tests working in DDIExporterTest

Fixed in 42985c6

I did not address the static/private suggestion for ExportService because @scolapasta wants to think about it some more.

I'm moving this to Code Review at https://waffle.io/IQSS/dataverse

The pull request is #3791.

**Comment by pdurbin** [comment]
Moving to QA based on @sekmiller 's "Looks good. Happy to see that the Exporter doesn't use/need the settings service."
